x/fungible/keeper/zrc20_cosmos_coins_mapping.go

```diff
@@ -0,0 +1,321 @@
+package keeper
+
+import (
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/zeta-chain/node/pkg/crypto"
+	fungibletypes "github.com/zeta-chain/node/x/fungible/types"
+)
+
+const (
+	transferFrom = "transferFrom"
+	transfer     = "transfer"
+	balanceOf    = "balanceOf"
+	allowance    = "allowance"
+)
+
+var (
+	ErrZRC20ZeroAddress    = fmt.Errorf("ZRC20 address cannot be zero")
+	ErrZRC20NotWhiteListed = fmt.Errorf("ZRC20 is not whitelisted")
+	ErrZRC20Paused         = fmt.Errorf("ZRC20 is paused")
+	ErrZRC20NilABI         = fmt.Errorf("ZRC20 ABI is nil")
+	ErrZeroAddress         = fmt.Errorf("address cannot be zero")
+	ErrInvalidAmount       = fmt.Errorf("amount must be positive")
+)
+
+// LockZRC20 locks ZRC20 tokens in the bank contract.
+// The caller must have approved the bank contract to spend the amount of ZRC20 tokens.
+func (k Keeper) LockZRC20(
+	ctx sdk.Context,
+	zrc20ABI *abi.ABI,
+	zrc20Address, from common.Address,
+	amount *big.Int,
+) error {
+	if zrc20ABI == nil {
+		return ErrZRC20NilABI
+	}
+
+	if amount.Sign() <= 0 || amount == nil {
+		return ErrInvalidAmount
+	}
+
+	if crypto.IsEmptyAddress(from) {
+		return ErrZeroAddress
+	}
+
+	if crypto.IsEmptyAddress(zrc20Address) {
+		return ErrZRC20ZeroAddress
+	}
+
+	if err := k.IsValidZRC20(ctx, zrc20Address); err != nil {
+		return err
+	}
+
+	if err := k.CheckFungibleZRC20Allowance(ctx, zrc20ABI, from, zrc20Address, amount); err != nil {
```

Comment:
Need some more comments in the function bodies, why this check is needed in this context for example. What allowance check allower, spender is being done